block/manager_test.go

```diff
@@ -535,42 +536,95 @@ func Test_publishBlock_ManagerNotProposer(t *testing.T) {
 	require.ErrorIs(err, ErrNotProposer)
 }
 
-func TestGetRemainingSleep(t *testing.T) {
-	now := time.Now()
-	interval := 10 * time.Second
-	defaultSleep := time.Second
-
-	// start height over the interval in the past
-	farPastStart := now.Add(-interval - 1)
-
-	// Table test in case we need to easily extend this in the future.
-	var tests = []struct {
+func TestManager_getRemainingSleep(t *testing.T) {
+	tests := []struct {
 		name          string
+		manager       *Manager
 		start         time.Time
-		interval      time.Duration
-		defaultSleep  time.Duration
 		expectedSleep time.Duration
 	}{
-		{"nil case", time.Time{}, 0, 0, 0},
-		{"start over the interval in the past", farPastStart, interval, defaultSleep, defaultSleep},
+		{
+			name: "Normal aggregation, elapsed < interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: false,
+				},
+			},
+			start:         time.Now().Add(-5 * time.Second),
+			expectedSleep: 5 * time.Second,
+		},
+		{
+			name: "Normal aggregation, elapsed >= interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: false,
+				},
+			},
+			start:         time.Now().Add(-15 * time.Second),
+			expectedSleep: 0,
+		},
+		{
+			name: "Lazy aggregation, building block, elapsed < interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: true,
+				},
+				buildingBlock: true,
+			},
+			start:         time.Now().Add(-5 * time.Second),
+			expectedSleep: 5 * time.Second,
+		},
+		{
+			name: "Lazy aggregation, building block, elapsed >= interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: true,
+				},
+				buildingBlock: true,
+			},
+			start:         time.Now().Add(-15 * time.Second),
+			expectedSleep: defaultLazyBufferTime,
+		},
+		{
+			name: "Lazy aggregation, not building block, elapsed < interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: true,
+				},
+				buildingBlock: false,
+			},
+			start:         time.Now().Add(-5 * time.Second),
+			expectedSleep: 15 * time.Second,
+		},
+		{
+			name: "Lazy aggregation, not building block, elapsed >= interval",
+			manager: &Manager{
+				conf: config.BlockManagerConfig{
+					BlockTime:      10 * time.Second,
+					LazyBlockTime:  20 * time.Second,
+					LazyAggregator: true,
+				},
+				buildingBlock: false,
+			},
+			start:         time.Now().Add(-25 * time.Second),
+			expectedSleep: 0,
+		},
```

Comment:
Can we get to here with block time being 0?

Reply:
Yeah, it's a lazy block and already exceed lazy block time, should produce block immediately